### ISSUE TITLE
Usage and signature of Model.train_step() is unclear

### ISSUE THREAD

**Issue by csxeba** [tensorflow/tensorflow#40402]
## URL(s) with the issue:

[Colab link](https://colab.research.google.com/drive/1A6gugIKUl3GCsGFlvmghFHQy5CAOd0RC?usp=sharing)

## Description of issue (what needs changing):

There seems to be some kind of restriction on the signature of Model.train_step(), which is undocumented.

### Clear description

The issue comes up in TF 2.2.0, where the possibility to overwrite Model.train_step() was introduced.

The training is executed with Model.fit() with a generator as an input. The generator yields 4 tensors, which are combined in train_step() to produce a loss and gradients.

The logic fails, because tf.keras in the fit() function checks that the generator outputs at most 3 tensors (corresponding to x, y, and weights), so this mindset and the x, y, w signature is implicity forced onto the train_step() function.

This is not clear from the documentation, which only states, that train_step() has a single argument (data), which should be a 'A nested structure of Tensors.' [Link to documentation](https://www.tensorflow.org/api_docs/python/tf/keras/Model#train_step)

The also seems to be a lack of official examples using this new API, which would help determine the sequence of methods one needs to call in order to use the new API (calling Model.compile() for instance).

### Expected behaviour

According to both the documentation and the release notes on TF 2.2.0, I was expecting to be able to use an arbitrary list of tensors as parameter to Model.train_step().

I would expect that a restriction on Model.train_step()'s sole 'data' argument would be documented in the relevant documentation.

## Disclaimer

Please note if I get the usage of Model.train_step() wrong. Also if there is an example out there using this new and very convenient API, feel free to direct me towards it.

**Comment by csxeba** [comment]
Update: I solved it by issuing the following modifications:

1) The data needs to be wrapped in a one-element tuple, eg. `yield ((tensor1, tensor2, ...),)`. This structure needs to be unpacked in the body of the train_step() function.

2) Added a line which builds the forward function. The line can either be a .predict() or a .build() call with the correct arguments.

It would be beneficial to somehow indicate in the documentation that explicit building of forward and backward passes are required for train_step() to work.

**Comment by nikitamaia** [comment]
Hi @csxeba, have you seen the [Customize what happens in Model.fit ](https://www.tensorflow.org/guide/keras/customizing_what_happens_in_fit) guide?
